crates/runner/src/lib.rs

```diff
@@ -60,8 +60,217 @@ pub struct RunnerOutput {
     pub public_address_ranges: PublicAddressRanges,
 }
 
+/// Total number of *inline memory cells* we must allocate below the args area
+/// when passing arguments according to the call ABI (arrays as pointers).
+/// This accounts for arrays at any nesting depth. Each array contributes:
+///   size * (arg_slot_size(element) + inline_data_slots_for_args(element))
+fn inline_data_slots_for_args(ty: &AbiType) -> usize {
+    match ty {
+        AbiType::FixedSizeArray { element, size } => {
+            let el_arg = AbiType::call_slot_size(element);
+            let el_inline = inline_data_slots_for_args(element);
+            (*size as usize) * (el_arg + el_inline)
+        }
+        AbiType::Tuple(ts) => ts.iter().map(inline_data_slots_for_args).sum(),
+        AbiType::Struct { fields, .. } => fields
+            .iter()
+            .map(|(_, t)| inline_data_slots_for_args(t))
+            .sum(),
+        // Non-arrays do not require extra inline materialization.
+        _ => 0,
+    }
+}
+
+/// Decode a value from VM memory using the *call ABI* layout starting at `base`.
+/// Returns the decoded value and the number of M31 cells consumed in that memory region.
+/// Arrays consume 1 cell in their containing region (a pointer), while their contents are
+/// read by following the pointer and decoding recursively.
+fn decode_value_from_call_memory(ty: &AbiType, vm: &VM, base: M31) -> Result<(CairoMValue, usize)> {
+    match ty {
+        AbiType::Felt => {
+            let m = vm.memory.get_data(base)?;
+            Ok((CairoMValue::Felt(m), 1))
+        }
+        AbiType::Bool => {
+            let m = vm.memory.get_data(base)?;
+            let v = m.0;
+            if v != 0 && v != 1 {
+                return Err(AbiCodecError::TypeMismatch(format!(
+                    "Invalid boolean value: expected 0 or 1, got {}",
+                    v
+                ))
+                .into());
+            }
+            Ok((CairoMValue::Bool(v == 1), 1))
+        }
+        AbiType::U32 => {
+            let lo = vm.memory.get_data(base)?;
+            let hi = vm.memory.get_data(base + M31::from(1u32))?;
+            if lo.0 >= (1 << 16) || hi.0 >= (1 << 16) {
```

Comment:
should constants be defined instead of magic numbers?
There might have other instances of this in the codebase. Maybe
```
const U16_MAX: u32 = 0xFFFF
if lo.0 > U16_MAX || hi.0 > U16_MAX
```

Reply:
it is obvious but for a reader there is a mental burden each time you encounter magic numbers you need to interpret. 
Anyway this is a nit.